### DISCUSSION TITLE
End of lesson 7(Interations.s.sol and InteractionTests.t.sol)  is incredibly hard

### DISCUSSION THREAD

**Discussion by elec7rik** [Cyfrin/foundry-full-course-cu#802] [General]
I can't understand it at all. Patrick is editing the code at light speed and the video seems to be cut and discontinued sometimes. How did you do that part? I need serious help with this it just seems not doable to me.,

**Comment by theirrationalone** [comment]
Hey @elec7rik!

Even after beholding carefully that excerpt, it's quite challenging to decode initially. 😩

However, through repetition, I managed to grasp it. Let me explain that if i can 😅... 

The 'Interactions' file is for automating interactions with contracts, such as funding and withdrawing in one go. Let's consider the 'FundfundMe' contract. To execute funding directly into the 'FundMe' contract, which was recently deployed, we used the 'run' function designed specifically for this task.

First, we obtained our recently deployed 'FundMe' contract using the 'foundry-devops' tool. Here, it's crucial to remember that if we haven't deployed our 'FundMe' contract with the '--broadcast' command, 'foundry-devops' will fail to retrieve the recently deployed 'FundMe' contract, resulting in the failure of our 'FundFundMe' script.

Patrick created the 'fundFundMe' function with a lowercase 'f' for simplicity. This allows us to call the 'fund' function not only from a recently deployed 'FundMe' contract but also from any instance of a 'FundMe' contract, regardless of when we deployed those instances.

Since we're funding, we also need to withdraw (typically by the owner). To follow the same strategy, we created the 'WithdrawFundMe' contract.

One interesting aspect of these scripts is that, undoubtedly, we're using 'foundry-devops' to retrieve our recently deployed 'FundMe' contract from anywhere.

These scripts are most helpful when we want to fund and withdraw in a single transaction. How? Let's say we first create/deploy and call 'FundFundMe,' retrieve our most recent 'FundMe' instance, and perform funding. Done! Similarly, we create/deploy and call 'WithdrawFundMe,' retrieve the same instance we previously retrieved in 'FundFundMe' (it's the most recent), and now we can perform withdrawals through it because we've already funded through the same instance.

Now, our new function can also be considered an integration test because it involves two transactions.

These are some of the most famous concepts in the programming world:

1. **DRY (Don't Repeat Yourself):** Avoiding code duplication.
2. **Outsourcing:** Delegating work and breaking tasks into smaller subtasks to avoid repetition.
3. **Encapsulation:** Creating functions that contain smaller functions to achieve the intended outcome.

If you find it beneficial for your workflow, you can further break down these scripts into smaller chunks.

Things to always keep in mind:

1. Who is calling whom? `vm.broadcast` in scripting corresponds to `vm.prank` in testing. 'Broadcast' requires the signer's private key (`vm.startBroadcast(pvt_key)`), while 'prank' uses the signer's address.

2. How will you maintain scripts to work on any chain? Use 'if' checks to compare chain IDs.

3. Ownership: Who owns the current transaction.
4. '--broadcast' flag for deployment.

**Reply by elec7rik** [reply]
thanks @theirrationalone , ig i will copy paste the code from github for these interactions files bc i still don't think i can write it down along with the videos. the tab switching and cut outs are just too weird. i will try to understand the code up to the best extent from your explanation and my understanding of the video.

**Reply by theirrationalone** [reply]
Not only you, Lots of us including me, followed the same strategy ∆... 

Good Luck 🍀 @elec7rik !

**Reply by cell0901** [reply]
how long have you been into solidity? can u share your LinkedIn?

**Reply by elec7rik** [reply]
@theirrationalone hi i got a this doubt if you could help me as you have finished the course so you know the flow of files: 
so i copied the Interactions.s.sol and InteractionsTest.t.sol and in the latter file there was a change- in the ```function setUp()``` there is the line 
```DeployFundMe deployer = new DeployFundMe();
       (fundMe, helperConfig) = deployer.run();
```
 whereas in the yt tutorial it is: 
```
DeployFundMe deploy = new DeployFundMe();
       fundMe = deploy.run();
```
the first one above(which i took from the current repo) throws the error: "Type contract FundMe is not implicitly convertible to expected type tuple(contract FundMe,contract HelperConfig)."
i changed it to the second one as it was in the video and it passes the tests without any error. are there any modifications in the current repo since the videos were released? 
also there is a difference in the function name: previously it was testUserCanFundInteractions() and now it is testUserCanFundAndOwnerWithdraw()
should i push it to github if it's passing all the tests? i don't see any obvious and apparent problem in the code.

**Reply by theirrationalone** [reply]
Thanks for calling me again @elec7rik ! 

As you know, Doing a same work again and again, can't be considered a smart work. 

In the deploy fundme script, we have an instance of helperConfig. Since we're importing that script into our fundme test script, So we don't need to deploy our fundme contract again therefore we already did outsourced that work into deploy fundme script. Right?!

Now take a look at that fundme deploy script once again. There we have returned our deployed fundme instance. By doing this, having this strategy, we don't have to deploy fundme contract again and again. Whenever we need to deploy fundme contract, we simply import & call deploy fundme script and unsurprisingly it returns an instance of fundme contract which is our ultimate need. 

Similarly, we know, we also have an instance of helperConfig into our deploy fundme script. So why not to return that also. By doing this we don't have to deploy helperConfig into our fundme test script again to work with it.

Common issues we could encounter if we don't do this. Creating a new instance of helperConfig can create new instances of mocks we implemented into `getOrCreateAnvilConfig`. But we have deployed our fundme contract with different instances of our mocks. 

Effectively, When we'll compare our `vrfcoordinator` address with a newly deployed helperConfig instance's `vrfcoordinator` address, there we can have a assertion error lvalue != rvalue. 


At this point, Hopefully you got the idea, what you need to do. 

Hint: test script expects that deploy script will return a tuple containing two values. But deploy script returns a tuple containing only one value.

Now, @elec7rik ! It is upto you what and how many values you want from deploy script. 

Good Luck 🍀

**Reply by elec7rik** [reply]
hi @theirrationalone thanks a lot bud your explanations are just amazing.
i just changed the return type by adding Helperconfig in the return type and return it as a tuple, i also changed in the fundMeTest.t.sol -> setUp() function. after that it is passing all the tests. should i now push it? also the code in the repo looks a bit different but the basic code structure is the same.

**Reply by theirrationalone** [reply]
Of course, @elec7rik !, You should push all of your code, even whenever you start any project which you mind to be on github. You should make a repo and should push your all phases of project i.e., initial project structure, scripts update, helpers for ease, tests, final push. As a result, It'll make your github profile amazing and you can show world, your proof of work. 😅

**Reply by elec7rik** [reply]
done, thanks bud.

**Reply by shrutic11** [reply]
Hey! I know how you got confused about Interactions. I spent more than 24 hours to make sure I capture I understand what's happening there, and then also help those who didn't understand it. If anyone is getting stuck with interactions and is eager to learn how it is working, please watch this explained video: https://youtu.be/4cuT8NTRwWE

I will also find some time to prepare an explainer article for this.

**Comment by 0xwtfwtf** [comment]
I am completely new to coding and I am now on lesson 12. There were definitely parts of the past projects that were really hard and didn't always make sense (although I always ended up being able to get them to work but it sometimes required posting a lot on here and using Google). However now that I am on lesson 12, things are starting to make more sense and I am able to code some things before we do them in the video. I realized that I probably need to go back and redo some of the interactions stuff and helperconfig stuff...like basically the stuff that isn't the actual contracts and writing unit tests. So I think what you're struggling with is harder but just keep going

**Reply by elec7rik** [reply]
you rocking! i am feeling better now that i've done that part, but it has taken a lot of time because sometimes it getes demoralising hard lol. but ya i'm doing it, just need to be more consistent. thanks and best luck to you too.
